src/core/qgstracer.cpp

```diff
@@ -0,0 +1,676 @@
+/***************************************************************************
+  qgstracer.cpp
+  --------------------------------------
+  Date                 : January 2016
+  Copyright            : (C) 2016 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstracer.h"
+
+#include "qgsgeometry.h"
+#include "qgsgeometryutils.h"
+#include "qgslogger.h"
+#include "qgsvectorlayer.h"
+
+#include <queue>
+#include <vector>
+
+typedef std::pair<int, double> queue_item; // first = vertex index, second = distance
+
+// utility comparator for queue items based on distance
+struct comp
+{
+  bool operator()( const queue_item &a, const queue_item &b )
+  {
+    return a.second > b.second;
+  }
+};
+
+
+// TODO: move to geometry utils
+double distance_2d( const QgsPolyline& coords )
```

Comment:
should be distance2D to keep consistent function naming (also closest_segment -> closestSegment )